### ISSUE TITLE
Che Tasks don't notify TaskManager when done if browser is closed

### ISSUE THREAD

**Issue by TheRealChssAddct** [eclipse-che/che#19612]
### Describe the bug

If the frontend (browser) is closed, a currently running Che Task's child process onClose event does not wind it's way  to the theia ide container hosting the backend Contributions, and thus the TaskManager tasks collection doesn't get updated (the finished task is never deleted).

### Che version

7.21.0 (da91724)

### Steps to reproduce

1. Create a che-theia extension with a backend contribution which injects TaskManager and WorkspaceService with essentially this code:
``` javascript

import { TaskManager } from '@theia/task/lib/node';
import { WorkspaceService } from '@eclipse-che/theia-remote-api/lib/common/workspace-service';

....

private startTimer(): void {
    setInterval(() => this.updateWorkspaceActivity(), TaskTrackerServiceImpl.REQUEST_PERIOD_MS);
}

private updateWorkspaceActivity(): void {
    const tasks = this.taskManager.getTasks(this.context);
    this.logger.info(`Task-tracker: Number of tasks: ${tasks?.length}`);
    const isAnyTaskRunning = !!tasks?.length;
    if (isAnyTaskRunning) {
        this.workspaceService.updateWorkspaceActivity(); // bump the idle timeout, keeping the workspace alive
    }
}
```
2. Build, deploy, create/run a workspace with the new extension.
3. Create a task to run something like this (importantly, must be type "che"):
``` json
{
    "type": "che",
    "label": "sleep3m",
    "command": "sleep 3m; echo 'done' > Sleep3mDone.txt;",
    "target": {
        "component": "my-container"
    },
    "problemMatcher": []
}
```
4. Test with keeping the browser open until the task completes.
   Observe task-tracker output in the theia ide container (k logs <pod> -c <container>) demonstrates the tasks count increase when the task starts, and decrease when the task finishes. 
   **Works as expected.**
6. Now start the task again, and then close your browser tab.
   Observe task-tracker output in the theia ide container demonstrates the tasks count increase when the task starts, but never decrease when the task finishes. 
   **Not working as expected.**

### Expected behavior

When the task finishes, regardless of the browser having been closed, the onClose event should reach the theia ide container, causing TaskManager to delete the task from it's collection of tasks.

### Runtime
   - [x] kubernetes 

Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:28:09Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}

Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"14f897abdc7b57f0850da68bd5959c9ee14ce2fe", GitTreeState:"clean", BuildDate:"2021-01-22T17:29:38Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}

### Installation method
   - [x] chectl

### Environment
   - [X] my computer
       - [x] Linux
   - [x] Cloud
       - [x] Azure

### Additional context

I originally created an extension using the pattern of eclipse-che-theia-activity-tracker to be a task tracker which keeps the workspace alive so long as a task is running, by bumping forward the idle timeout. A frontendContribution which injects TaskWatcher and TaskService handles taskWatcher.onTaskCreated() and taskWatcher.onTaskExit() events and uses taskService.getRunningTasks() to determine if there are active jobs. In a timer loop while there are active jobs, it signals a backendContribution which sends a request to workspace master using workspaceService.updateWorkspaceActivity(). This is all working as expected. (thanks to @vzhukovskyi for pointing me in the right direction).

Because it uses a frontend contribution, it requires the client-side browser to remain open so I added the onWillStop() method, which at least warns the user when they attempt to close the browser (or browser tab), (although the message text is not able to be set, so it's not an entirely helpful warning).

I wanted to create an entirely server-side solution, so the user could close their browser and the task tracker would continue to function correctly, so that as long as the task was running, the workspace would not shut down. I modified the backendContribution to simply enter an infinite loop, which then checks taskManager.getTasks() to determine if there are active tasks and if so bumps the idle timeout. In my testing, this works. For my testing, I was using tasks of type="shell", using the sleep command. 

Once it got to QA, they tested using tasks of type="che", with a target container. This is where the problem appears. If the browser is closed, the task does continue to run and finish in it's target container (this is good), but when the task finishes the onClose event doesn't ever reach the backendContribution which is hosted in the theia ide container (not good). I got an interactive shell to the container running the task to determine the task does indeed finish. (I also tested by echoing output to a file). But the taskManager never deletes the finished task from it's tasks Map (it originally added the task when the task was started while the browser was still open). 

Therefore, the task tracker continues infinitely to bump the idle timeout, effectively keeping the workspace pod/containers alive forever, sitting idle.

**Comment by RomanNikitenko** [comment]
Thanks for the raising the issue, we need investigate it.

I'm adding the comment from @tsmaeder from mattermost channel:

> Theia starts a connection module for each front end that connects to the back end. I suspect when your browser closes, the back end connection module gets stopped as well.

**Comment by akurinnoy** [comment]
@TheRealChssAddct can you check your issue with a fresher Che (the latest available 7.28)?

**Comment by TheRealChssAddct** [comment]
> @TheRealChssAddct can you check your issue with a fresher Che (the latest available 7.28)?

Hehe. That's fair. I expected that.

I anticipate that in May we will be doing a catch-up update to the freshest version.

Once our update is complete, this is something I will revisit, as it is very much wanted, and then comment here with status.

--- 

As an aside, I will say I really appreciate how active/aggressive this project is!  But what it does mean is that if we get too far behind, it makes updating to freshest a bigger deal. Last one things were split into separate repos. In starting to scope out the next update, i see there's a frontend tech change (angular to react, I believe), other repo changes, I see that Multi-Root will now be the only option, so we'lI totally miss the 'gradual change' efforts and have to go all the way from A-Z. 

But I much prefer all this activity to a slow, plodding project.  So, kudos to y'all!
